wallet, rpc: add v3 transaction creation and wallet support

ishaanam opened the pull request:
This PR Implements the following:
- If creating a v3 transaction, `AvailableCoins` doesn't return unconfirmed v2 utxos (and vice versa)
- `AvailableCoins` doesn't return an unconfirmed v3 utxo if its transaction already has a child
- If a v3 transaction is kicked out of the mempool by a sibling, mark the sibling as a mempool conflict
- Throw an error if pre-selected inputs are of the wrong transaction version
- Allow setting version to 3 manually in `createrawtransaction` (uses commits from #31936)
- Limits a v3 transaction weight in coin selection

Closes #31348

To-Do:
- [x] Test a v3 sibling conflict kicking out one of our transactions from the mempool
- [x] Implement separate size limit for TRUC children
- [x] Test that we can't fund a v2 transaction when everything is v3 unconfirmed
- [x] Test a v3 sibling conflict being removed from the mempool
- [x] Test limiting v3 transaction weight in coin selection
- [x] Simplify tests
- [x] Add documentation 
- [x] Test that user-input max weight is not overwritten by truc max weight
- [x] Test v3 in RPCs other than `createrawtransaction`

Bue-von-hon committed:
rpc: Add transaction min standard version parameter

Bue-von-hon committed:
rpc: Support version 3 transaction creation

Adds v3 support to the following RPCs:
- createrawtransaction
- createpsbt
- send
- sendall
- walletcreatefundedpsbt

Co-authored-by: chungeun-choi <cucuridas@gmail.com>
Co-authored-by: dongwook-chan <dongwook.chan@gmail.com>
Co-authored-by: sean-k1 <uhs2000@naver.com>
Co-authored-by: ishaanam <ishaana.misra@gmail.com>

ishaanam committed:
wallet: unconfirmed ancestors and descendants are always truc

ishaanam committed:
wallet: limit v3 tx weight in coin selection

ishaanam committed:
wallet: mark unconfirmed v3 siblings as mempool conflicts

glozow commented on src/wallet/rpc/spend.cpp:
```diff
@@ -715,6 +716,12 @@ CreatedTransactionResult FundTransaction(CWallet& wallet, const CMutableTransact
     }
 
+    if (tx.version == TRUC_VERSION) {
+        if (!coinControl.m_max_tx_weight.has_value() || coinControl.m_max_tx_weight.value() > 40000) {
```
Instead of magic numbers, try using `TRUC_MAX_VSIZE * WITNESS_SCALE_FACTOR`

glozow replied:
Also, there is a separate size limit for TRUC children - I don't think that's been implemented yet?

ishaanam replied:
Yes, I will look into implementing the separate size limit.

glozow commented on src/wallet/transaction.h:
```diff
@@ -258,6 +258,9 @@ class CWalletTx
     // BlockConflicted.
     std::set<Txid> mempool_conflicts;
 
+    // Set of v3 transactions that spend from this tx
```
Comment seems inaccurate: it mentions a set but this only allows for 1 transaction. Should also mention (1) that this is used to stop us from creating another unconfirmed child and (2) this is specifically the in mempool-sibling, as there can be multiple siblings but only 1 in mempool (unless there was a reorg).

ishaanam replied:
Done

glozow reviewed (commented):
Nice, concept ACK! Saw that these tests fail on #31936, which is helpful for showing its issues. Ultimately, I think the wallet commits should be introduced before the RPC ones.

ishaanam committed:
wallet: don't include unconfirmed v3 txs with children in available coins

ishaanam committed:
wallet: throw error at conflicting tx versions in pre-selected inputs

ishaanam commented on test/functional/wallet_v3_txs.py:
```diff
@@ -0,0 +1,111 @@
+        self.v3_utxos_appear_in_listunspent()
+
+    @cleanup
+    def v3_tx_spends_unconfirmed_v2_tx(self):
```
Good point, I will add a test for this.

ishaanam commented on src/rpc/rawtransaction_util.cpp:
```diff
@@ -154,6 +155,13 @@ CMutableTransaction ConstructTransaction(const UniValue& inputs_in, const UniVal
+    if (!version.isNull()) {
+        uint32_t nVersion = version.getInt<uint32_t>();
+        if (nVersion < TX_MIN_STANDARD_VERSION || nVersion > TX_MAX_STANDARD_VERSION)
+            throw JSONRPCError(RPC_INVALID_PARAMETER, std::string("Invalid parameter, version out of range(") + util::ToString(TX_MIN_STANDARD_VERSION) + "~" + util::ToString(TX_MAX_STANDARD_VERSION) + ")");
```
Done

rkrux reviewed (commented):
Concept ACK b3817a822054fb2ea2964535b5a0ee55dd5d55ac
Thanks for picking this up, I will try to review it soon.

The TODOs seem to be done and can be removed from the PR description.

ishaanam commented on src/rpc/rawtransaction_util.cpp:
```diff
@@ -154,6 +155,15 @@ CMutableTransaction ConstructTransaction(const UniValue& inputs_in, const UniVal
+    if (!version.isNull()) {
+        uint32_t nVersion = version.getInt<uint32_t>();
+        if (nVersion < TX_MIN_STANDARD_VERSION || nVersion > TX_MAX_STANDARD_VERSION) {
+            // throw JSONRPCError(RPC_INVALID_PARAMETER, std::string("Invalid parameter, version out of range(") + util::ToString(TX_MIN_STANDARD_VERSION) + "~" + util::ToString(TX_MAX_STANDARD_VERSION) + ")");
```
Done

ishaanam commented on src/rpc/rawtransaction.cpp:
```diff
@@ -437,7 +438,7 @@ static RPCHelpMan createrawtransaction()
         rbf = request.params[3].get_bool();
     }
-    CMutableTransaction rawTx = ConstructTransaction(request.params[0], request.params[1], request.params[2], rbf);
+    CMutableTransaction rawTx = ConstructTransaction(request.params[0], request.params[1], request.params[2], rbf, request.params[4]);
```
Done

ishaanam commented on src/rpc/rawtransaction.cpp:
```diff
@@ -437,7 +438,7 @@ static RPCHelpMan createrawtransaction()
         rbf = request.params[3].get_bool();
     }
-    CMutableTransaction rawTx = ConstructTransaction(request.params[0], request.params[1], request.params[2], rbf);
+    CMutableTransaction rawTx = ConstructTransaction(request.params[0], request.params[1], request.params[2], rbf, self.Arg<unsigned long>("version"));
```
this fixed the error, thanks!

ishaanam committed:
doc: add release notes for version 3 transactions

ishaanam commented on test/functional/wallet_v3_txs.py:
```diff
@@ -0,0 +1,396 @@
+        self.sync_mempools()
+        return txid
+
+    def bulk_tx(self, tx, amount, target_vsize):
```
I've added a `bulk_vout` function to `script_util.py` which is used by `MiniWallet` and these tests.

ishaanam commented on test/functional/wallet_v3_txs.py:
```diff
@@ -0,0 +1,396 @@
+        # alice spends her output with a v3 transaction
+        alice_unspent = self.alice.listunspent(minconf=0)[0]
+        inputs=[{'txid' : parent_txid, 'vout' : alice_unspent['vout']},]
+        outputs = {self.alice.getnewaddress() : alice_unspent['amount'] - Decimal(0.00000120)} # two outputs
```
Yes, this was an outdated comment which I've now removed

achow101 replied:
The comment still seems to be there.

ishaanam replied:
The comment was in multiple places, I've removed it now.

ishaanam commented on test/functional/wallet_v3_txs.py:
```diff
@@ -0,0 +1,396 @@
+        self.send_tx(self.bob, inputs, outputs, 3)
+
+    @cleanup
+    def mempool_conflicts_removed_when_v3_conflict_removed(self):
```
It would be tricky to consolidate the two because one of them is testing preventing the wallet from creating a transaction and the other is testing marking conflicts correctly.

ishaanam commented:
Thanks for the review @glozow! I've addressed your comments and also refactored a few of the functional tests.

murchandamus reviewed (commented):
Concept ACK

ishaanam committed:
wallet: set m_version in coin control to default value

In future commits we assume that coin_control.m_version has a
value when making sure that we follow truc rules, so we should
give it a default value of CTransaction::CURRENT_VERSION.

ishaanam commented on src/wallet/spend.cpp:
```diff
@@ -283,6 +283,16 @@ util::Result<PreSelectedInputs> FetchSelectedInputs(const CWallet& wallet, const
                 input_bytes = CalculateMaximumSignedInputSize(txout, &wallet, &coin_control);
             }
+            auto it = wallet.mapWallet.find(outpoint.hash);
+            if (coin_control.m_version.has_value() && it != wallet.mapWallet.end()) {
```
This is a good point, if `coin_control.m_version` is not set and the preselected inputs have different versions, then an error would not be thrown. I have changed `m_version` so that it has a default value of `CTransaction::CURRENT_VERSION` instead of being optional.

ishaanam commented on test/functional/wallet_v3_txs.py:
```diff
@@ -0,0 +1,399 @@
+        )
+
+    @cleanup
+    def spend_inputs_with_different_versions(self, version_a, version_b):
```
I've added a test for this.

glozow commented on src/wallet/rpc/spend.cpp:
```diff
@@ -1302,7 +1302,7 @@ RPCHelpMan send()
                     InterpretSubtractFeeFromOutputInstructions(options["subtract_fee_from_outputs"], outputs.getKeys())
             );
-            CMutableTransaction rawTx = ConstructTransaction(options["inputs"], request.params[0], options["locktime"], rbf);
+            CMutableTransaction rawTx = ConstructTransaction(options["inputs"], request.params[0], options["locktime"], rbf, DEFAULT_RAWTX_VERSION);
```
I think perhaps the wallet's default tx version should be specified in a separate constexpr that lives in the wallet code?

ishaanam replied:
Done

glozow commented on src/policy/policy.cpp:
```diff
@@ -98,7 +98,7 @@ bool IsStandard(const CScript& scriptPubKey, TxoutType& whichType)
 bool IsStandardTx(const CTransaction& tx, const std::optional<unsigned>& max_datacarrier_bytes, bool permit_bare_multisig, const CFeeRate& dust_relay_fee, std::string& reason)
 {
-    if (tx.version > TX_MAX_STANDARD_VERSION || tx.version < 1) {
+    if (tx.version > TX_MAX_STANDARD_VERSION || tx.version < TX_MIN_STANDARD_VERSION) {
```
This change is in 2bcdf53a58ee6cd7c70f2afd97a16fce5dcc21c6 but seems to belong in b07d9b92821 instead

ishaanam replied:
I've moved the change to the correct commit

glozow commented on test/functional/test_framework/mempool_util.py:
```diff
@@ -30,6 +30,11 @@
     MiniWallet,
 )
 
+ORPHAN_TX_EXPIRE_TIME = 1200
```
Orphans don't expire anymore and this is unused (is it from a rebase maybe?)

ishaanam replied:
I've removed it, pretty sure it was from a rebase

glozow commented on test/functional/test_framework/wallet.py:
```diff
@@ -121,17 +120,9 @@ def _bulk_tx(self, tx, target_vsize):
         """Pad a transaction with extra outputs until it reaches a target vsize.
         returns the tx
         """
-        if target_vsize < tx.get_vsize():
-            raise RuntimeError(f"target_vsize {target_vsize} is less than transaction virtual size {tx.get_vsize()}")
-
         tx.vout.append(CTxOut(nValue=0, scriptPubKey=CScript([OP_RETURN])))
-        # determine number of needed padding bytes
-        dummy_vbytes = target_vsize - tx.get_vsize()
-        # compensate for the increase of the compact-size encoded script length
-        # (note that the length encoding of the unpadded output script needs one byte)
-        dummy_vbytes -= len(ser_compact_size(dummy_vbytes)) - 1
-        tx.vout[-1].scriptPubKey = CScript([OP_RETURN] + [OP_1] * dummy_vbytes)
-        assert_equal(tx.get_vsize(), target_vsize)
```
Maybe put the `bulk_vout` refactor into a separate commit

glozow commented on test/functional/wallet_v3_txs.py:
```diff
@@ -0,0 +1,422 @@
+# Copyright (c) 2025 The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+"""Test how the wallet deals with v3 transactions"""
```
s/v3/TRUC

glozow reviewed (commented):
Looks pretty good! Generally, I prefer more detailed comments when the code is not 100% self-explanatory, particularly for the confusing areas like mempool conflicts

ishaanam commented on src/wallet/spend.cpp:
```diff
@@ -390,6 +390,7 @@ CoinsResult AvailableCoins(const CWallet& wallet,
                     if (wtx.tx->version != TRUC_VERSION) continue;
+                    if (wtx.v3_spend.has_value()) continue; // this unconfirmed v3 transaction already has a child
                 } else {
                     if (wtx.tx->version == TRUC_VERSION) continue;
```
Why is it safe to assume that the transaction doesn't already have a child here?

glozow replied:
non-TRUC transactions shouldn't have TRUC children. I think it would be a bug if we set this variable for a transaction that isn't TRUC

ishaanam replied:
Oh, that makes sense, I thought that you were talking about changing the if-statement on line 393.

ishaanam committed:
test: extract `bulk_vout` from `bulk_tx` so it can be used by wallet tests

ishaanam committed:
test: add truc wallet tests

ishaanam commented on src/wallet/wallet.cpp:
```diff
@@ -1446,6 +1462,19 @@ void CWallet::transactionRemovedFromMempool(const CTransactionRef& tx, MemPoolRe
         }
     }
+
+    // reverse what happens in transactionAddedToMempool
```
Done

ishaanam commented on src/wallet/coincontrol.h:
```diff
@@ -109,10 +109,10 @@ class CCoinControl
     //! Locktime
     std::optional<uint32_t> m_locktime;
-    //! Version
-    std::optional<uint32_t> m_version;
```
Added description to commit message.

ishaanam commented on src/wallet/wallet.cpp:
```diff
@@ -1406,6 +1406,17 @@ void CWallet::transactionAddedToMempool(const CTransactionRef& tx) {
+                    // Find all wallet transactions that spend utxos from this tx
+                    for (long unsigned int i = 0; i < wtx.tx->vout.size(); i++) {
+                        for (auto range = mapTxSpends.equal_range(COutPoint(wtx.tx->GetHash(), i)); range.first != range.second; range.first++) {
+                            const Txid& spent_id = range.first->second;
```
Done

ishaanam commented on src/wallet/wallet.cpp:
```diff
@@ -1393,6 +1394,21 @@ void CWallet::transactionAddedToMempool(const CTransactionRef& tx) {
+                CWalletTx& wtx = wallet_it->second;
+                if (wtx.isUnconfirmed()) {
+                    wtx.v3_spend = tx->GetHash();
+                }
```
Done

glozow commented on src/wallet/wallet.cpp:
```diff
@@ -1475,6 +1489,15 @@ void CWallet::transactionRemovedFromMempool(const CTransactionRef& tx, MemPoolRe
                 CWalletTx& wtx = wallet_it->second;
                 if (wtx.truc_child_in_mempool == tx->GetHash()) {
                     wtx.truc_child_in_mempool = std::nullopt;
+                    // Find all wallet transactions that spend utxos from this tx
```
This comment can be confusing - we're not actually looking for transactions spending from *this* tx. We're looking for siblings of this tx, which spend from utxos of `wtx` (which is actually this transaction's parent). Might be helpful to rename `wallet_it` to `parent_it` and update this comment.

ishaanam replied:
Done

glozow reviewed (commented):
a2c43c3b9447377d2f2635e23c9043e40b769dd4 looks pretty good to me! Minor comments mostly about docs

ishaanam commented on src/wallet/spend.h:
```diff
@@ -83,6 +83,8 @@ struct CoinFilterParams {
     // By default, skip locked UTXOs
     bool skip_locked{true};
+    // Whether or not to care about the tx version
+    bool track_version{true}; // only used by AvailableCoinsListUnspent
```
Done

ishaanam commented on src/wallet/spend.cpp:
```diff
@@ -390,8 +390,10 @@ CoinsResult AvailableCoins(const CWallet& wallet,
             if (nDepth == 0 && params.track_version) {
                 if (coinControl->m_version == TRUC_VERSION) {
                     if (wtx.tx->version != TRUC_VERSION) continue;
+                    if (wtx.truc_child_in_mempool.has_value()) continue; // this unconfirmed v3 transaction already has a child
```
Done

ishaanam commented on src/wallet/wallet.cpp:
```diff
@@ -1446,6 +1463,22 @@ void CWallet::transactionRemovedFromMempool(const CTransactionRef& tx, MemPoolRe
+
+    if (tx->version == TRUC_VERSION) {
+        // unset truc_child_in_mempool to make it possible to spend from
+        // this again. If this tx was replaced by another
```
Done

ishaanam commented on src/rpc/rawtransaction_util.h:
```diff
@@ -53,6 +53,6 @@ std::vector<std::pair<CTxDestination, CAmount>> ParseOutputs(const UniValue& out
 
 /** Create a transaction from univalue parameters */
-CMutableTransaction ConstructTransaction(const UniValue& inputs_in, const UniValue& outputs_in, const UniValue& locktime, std::optional<bool> rbf);
+CMutableTransaction ConstructTransaction(const UniValue& inputs_in, const UniValue& outputs_in, const UniValue& locktime, std::optional<bool> rbf, const uint32_t& version);
```
Done

ishaanam commented on doc/release-notes-32896.md:
```diff
@@ -0,0 +1,19 @@
+Wallet
+------
+Support has been added for spending version 3 transactions received by
+the wallet. The wallet ensures that version 3 policy rules are being
```
Done

ishaanam commented on src/wallet/spend.cpp:
```diff
@@ -924,11 +924,17 @@ util::Result<SelectionResult> AutomaticCoinSelection(const CWallet& wallet, Coin
             auto it = filtered_groups.find(select_filter.filter);
             if (it == filtered_groups.end()) continue;
+            if (temp_selection_params.m_version == TRUC_VERSION && (select_filter.filter.conf_mine == 0 || select_filter.filter.conf_theirs == 0)) {
+                if (temp_selection_params.m_max_tx_weight > (TRUC_CHILD_MAX_VSIZE * WITNESS_SCALE_FACTOR)) {
+                    temp_selection_params.m_max_tx_weight = TRUC_CHILD_MAX_VSIZE * WITNESS_SCALE_FACTOR;
```
I've added an additional test case.

achow101 commented on test/functional/wallet_v3_txs.py:
```diff
@@ -0,0 +1,560 @@
+
+        inputs=[]
+        outputs = {self.alice.getnewaddress() : 2.0, self.bob.getnewaddress() : 2.0}
+        self.send_tx(self.charlie, inputs, outputs, 3)
+        parent_txid = self.charlie.getrawmempool()[0]
```
In 09c9f828250e3840d7557374ff568204305e2679 "test: add truc wallet tests"

The `getrawmempool` can be avoided:

```suggestion
        parent_txid = self.send_tx(self.charlie, inputs, outputs, 3)
```

achow101 commented on test/functional/wallet_v3_txs.py:
```diff
@@ -0,0 +1,560 @@
+def cleanup(func):
+    def wrapper(self, *args):
+        try:
+            func(self, *args)
```
In 09c9f828250e3840d7557374ff568204305e2679 "test: add truc wallet tests"

Since all of the test cases do a `self.generate`, that could be added here as well before `func`.

achow101 commented on test/functional/wallet_v3_txs.py:
```diff
@@ -0,0 +1,560 @@
+        )
+
+    @cleanup
+    def truc_tx_with_conflicting_sibling_change(self):
```
In 09c9f828250e3840d7557374ff568204305e2679 "test: add truc wallet tests"

I don't see how this test is materially different from `truc_tx_with_conflicting_sibling`. Based on the naming, I assume this was added in response to https://github.com/bitcoin/bitcoin/pull/32896/commits/09c9f828250e3840d7557374ff568204305e2679#r2211372152? However, this test isn't testing anything with a change output since the parent tx is from charlie.

ishaanam replied:
I've updated this test so it actually tests spending change now.

achow101 commented on test/functional/wallet_v3_txs.py:
```diff
@@ -0,0 +1,560 @@
+        alice_unspent = [unspent for unspent in self.alice.listunspent(minconf=0) if unspent["confirmations"] == 0][0]
+
+        # alice spends both of her outputs
+        inputs = [{'txid' : alice_v2_unspent['txid'], 'vout' : 0}, {'txid' : alice_unspent['txid'], 'vout' : alice_unspent['vout']}]
```
In 09c9f828250e3840d7557374ff568204305e2679 "test: add truc wallet tests"

For correctness, use `alive_v2_unspent["vout"]` rather than assume the output is at 0.

ishaanam replied:
Done

achow101 commented on test/functional/wallet_v3_txs.py:
```diff
@@ -0,0 +1,560 @@
+
+        # alice spends her output with a v3 transaction
+        alice_unspent = self.alice.listunspent(minconf=0)[0]
+        inputs=[{'txid' : parent_txid, 'vout' : alice_unspent['vout']},]
```
In 09c9f828250e3840d7557374ff568204305e2679 "test: add truc wallet tests"

Since we already have the UTXO from `listunspent`, we don't need to specify the txid and vout individually.

```suggestion
        inputs=[alice_unspent]
```

Since this is short, it could also be inlined. This comment applies throughout this file.

ishaanam replied:
Done here and throughout the file

achow101 commented on test/functional/wallet_v3_txs.py:
```diff
@@ -0,0 +1,560 @@
+        self.send_tx(self.charlie, inputs, outputs, 3)
+
+        alice_v2_unspent = self.alice.listunspent(minconf=1)[0]
+        alice_unspent = [unspent for unspent in self.alice.listunspent(minconf=0) if unspent["confirmations"] == 0][0]
```
In 09c9f828250e3840d7557374ff568204305e2679 "test: add truc wallet tests"

`listunspent` also has a `maxconf`, so this could be `self.alice.listunspent(minconf=0, maxconf=0)[0]`

achow101 commented on test/functional/wallet_v3_txs.py:
```diff
@@ -0,0 +1,560 @@
+    @cleanup
+    def walletcreatefundedpsbt_v3(self):
+        self.log.info("Test setting version to 3 with walletcreatefundedpsbt")
+        self.log.info("Test setting version to 3 with createpsbt")
```
In 09c9f828250e3840d7557374ff568204305e2679 "test: add truc wallet tests"

This log is not supposed to be here.

ishaanam replied:
I've removed it.

achow101 reviewed (commented):
I think `sendall` is missing the weight check if the transaction is truc.

ishaanam commented on src/wallet/wallet.cpp:
```diff
@@ -1404,9 +1404,23 @@ void CWallet::transactionAddedToMempool(const CTransactionRef& tx) {
-                    parent_wtx.truc_child_in_mempool = tx->GetHash();
+                CWalletTx& parent_tx = parent_it->second;
+                if (parent_tx.isUnconfirmed()) {
+                    parent_tx.truc_child_in_mempool = tx->GetHash();
```
I've fixed how these variables are named.

ishaanam commented on src/wallet/rpc/spend.cpp:
```diff
@@ -1303,6 +1310,15 @@ RPCHelpMan send()
+                } else {
+                    coin_control.m_max_tx_weight = TRUC_MAX_VSIZE * WITNESS_SCALE_FACTOR;
+                }
+            }
```
I've removed it.

ishaanam commented on test/functional/wallet_v3_txs.py:
```diff
@@ -0,0 +1,560 @@
+
+    def run_test(self):
+        self.connect_nodes(0, 1)
+        self.connect_nodes(0, 2)
```
I've removed it

ishaanam commented on test/functional/wallet_v3_txs.py:
```diff
@@ -0,0 +1,560 @@
+
+    def set_test_params(self):
+        getcontext().prec=10
+        self.num_nodes = 3
```
Done

ishaanam commented on test/functional/wallet_v3_txs.py:
```diff
@@ -0,0 +1,560 @@
+
+    @cleanup
+    def v2_tx_spends_confirmed_v3_tx(self, version_a, version_b):
+        self.log.info(f"Test unavailable funds when v{version_a} tx spends confirmed v{version_b} tx")
```
Done

w0xlt commented on src/wallet/coinselection.h:
```diff
@@ -174,6 +174,8 @@ struct CoinSelectionParams {
      */
     bool m_include_unsafe_inputs = false;
+    /** The version of the transaction we are trying to create. */
+    uint32_t m_version = CTransaction::CURRENT_VERSION;
```
nit
```suggestion
    uint32_t m_version{CTransaction::CURRENT_VERSION};
```

ishaanam replied:
Done

glozow reviewed (commented):
ACK 43a479ca688, only minor nits

w0xlt reviewed (commented):
ACK https://github.com/bitcoin/bitcoin/pull/32896/commits/43a479ca6885b4222fb5e0f673354e54144d6835

ishaanam commented on test/functional/wallet_v3_txs.py:
```diff
@@ -0,0 +1,560 @@
+
+        assert_equal(decoded_tx["version"], 3)
+
+        assert confirmed_v3 in decoded_vin_txids
```
Thanks for finding this! I've fixed it by explicitly choosing the utxos to spend. I've also tested that this works by setting a lower feerate in `send_tx` and ensuring that the test still passes.

ishaanam commented on src/wallet/rpc/spend.cpp:
```diff
@@ -1483,6 +1504,11 @@ RPCHelpMan sendall()
                         continue;
                     }
+                    // we are spending a TRUC child, so we have a
+                    // lower max weight
```
Done

ishaanam commented on test/functional/wallet_v3_txs.py:
```diff
@@ -0,0 +1,548 @@
+        )
+
+    @cleanup
+    def v2_tx_spends_confirmed_v3_tx(self, version_a, version_b):
```
Done

ishaanam commented on test/functional/wallet_v3_txs.py:
```diff
@@ -0,0 +1,548 @@
+
+    @cleanup
+    def send_v3(self):
+        self.log.info("Test setting version to 3 with `send`")
```
Done

glozow reviewed (commented):
reACK 013cbf7eda2

Changes were braced initialization, docs/rename nits, and changing `sendall_with_unconfirmed_v3` to choose UTXOs explicitly so we can assert what's spent at the end.

achow101 commented on test/functional/wallet_v3_txs.py:
```diff
@@ -0,0 +1,553 @@
+        self.send_tx(self.charlie, [], outputs, 2)
+        self.generate(self.nodes[0], 1)
+
+        unspent1 = self.alice.listunspent()[0]
+        unspent2 = self.alice.listunspent()[1]
+        unspent3 = self.alice.listunspent()[2]
+        unspent4 = self.alice.listunspent()[3]
```
In f4d209c0e94fad8435215abba3e61d79585d0d12 "test: add truc wallet tests"

Could store the result of `listunspent` and access the specific utxos by index later:

```suggestion
        utxos = self.alice.listunspent()
```

ishaanam replied:
Done

achow101 commented on test/functional/wallet_v3_txs.py:
```diff
@@ -0,0 +1,553 @@
+    def sendall_with_unconfirmed_v3(self):
+        self.log.info("Test setting version to 3 with sendall + unconfirmed inputs")
+
+        outputs = {}
+        for _ in range(4):
+            outputs[self.alice.getnewaddress()] = 2.00001
```
In f4d209c0e94fad8435215abba3e61d79585d0d12 "test: add truc wallet tests"

This could be condensed into a one-liner

```suggestion
        outputs = {self.alice.getnewaddress(): 2.00001 for _ in range(4)}
```

ishaanam replied:
Done

achow101 reviewed (commented):
ACK 013cbf7eda2a3c643668b467cdbeb37027044b3a

w0xlt reviewed (commented):
ACK https://github.com/bitcoin/bitcoin/pull/32896/commits/013cbf7eda2a3c643668b467cdbeb37027044b3a

TRUC rules implemented:  

| Rules | Commits |
|---|:---:|
| 1. A TRUC tx must only have TRUC unconfirmed ancestors | afbf6c1, f3aba33 |
| 2. A non‑TRUC tx must only have non‑TRUC unconfirmed ancestors | afbf6c1; f3aba33 |
| 3. A TRUC’s ancestor set (incl. itself) must be within `TRUC_ANCESTOR_LIMIT` | Enforced by mempool/TRUC policy (outside this PR) |
| 4. A TRUC’s descendant set (incl. itself) must be within TRUC_DESCENDANT_LIMIT | 1ce3c06, 9921f22 |
| 5. If a TRUC tx has any unconfirmed ancestors, the tx’s sigop‑adjusted vsize must be ≤ TRUC_CHILD_MAX_VSIZE | eca8f1f | 
| 6. A TRUC tx must be ≤ TRUC_MAX_VSIZE | eca8f1f |

All of these rules (except the third one) are included in the added test file.

rkrux commented on test/functional/test_framework/script_util.py:
```diff
@@ -128,6 +131,16 @@ def script_to_p2sh_p2wsh_script(script):
+    # compensate for the increase of the compact-size encoded script length
+    # (note that the length encoding of the unpadded output script needs one byte)
+    dummy_vbytes -= len(ser_compact_size(dummy_vbytes)) - 1
+    tx.vout[-1].scriptPubKey = CScript([OP_RETURN] + [OP_1] * dummy_vbytes)
```
In 633aa947d684a0fdd3c13353ddd8418cb0195aa5 "test: extract `bulk_vout` from `bulk_tx` so it can be used by wallet tests"

Fine to extract it but since it particularly bulks the last `vout`, we can call the function `bulk_last_vout`? An alternative could be that this function can append the op_return vout in the transaction instead of assuming the caller does it, in which case this function would become `bulk_tx` - but this could be a larger change and can be avoided in this PR.

rkrux commented on test/functional/wallet_v3_txs.py:
```diff
@@ -0,0 +1,553 @@
+            "Maximum transaction weight is less than transaction weight without inputs",
+            self.charlie.fundrawtransaction,
+            tx.serialize_with_witness().hex(),
+            {'include_unsafe' : True, 'max_tx_weight' : int(TRUC_MAX_VSIZE/2)}
```
In f4d209c0e94fad8435215abba3e61d79585d0d12 "test: add truc wallet tests"

The user-input tx weight is the same as the previously bulk-ed tx weight?

rkrux commented on test/functional/wallet_v3_txs.py:
```diff
@@ -0,0 +1,553 @@
+        getcontext().prec=10
+        self.num_nodes = 1
+        self.setup_clean_chain = True
+        # whitelist peers to speed up tx relay / mempool sync
+        self.noban_tx_relay = True
```
In https://github.com/bitcoin/bitcoin/commit/f4d209c0e94fad8435215abba3e61d79585d0d12 "test: add truc wallet tests"

I don't think this is required because there is only one node in the test.

rkrux commented on test/functional/wallet_v3_txs.py:
```diff
@@ -0,0 +1,553 @@
+    @cleanup
+    def truc_tx_with_conflicting_sibling_change(self):
+        # unconfirmed v3 tx to alice & bob
+        self.log.info("Test v3 transaction with conflicting sibling")
```
To deduplicate subtest logs.

```diff
- self.log.info("Test v3 transaction with conflicting sibling")
+ self.log.info("Test v3 transaction with conflicting sibling change") 
```

ishaanam replied:
Done

rkrux reviewed (commented):
Looking good, code review 013cbf7eda2a3c643668b467cdbeb37027044b3a

I have shared few suggestions, all are minor I believe.

rkrux commented on src/wallet/spend.cpp:
```diff
@@ -283,6 +283,14 @@ util::Result<PreSelectedInputs> FetchSelectedInputs(const CWallet& wallet, const
                 input_bytes = CalculateMaximumSignedInputSize(txout, &wallet, &coin_control);
             }
+            const CWalletTx& parent_tx = txo->GetWalletTx();
+            if (wallet.GetTxDepthInMainChain(parent_tx) == 0) {
+                if (parent_tx.tx->version == TRUC_VERSION && coin_control.m_version != TRUC_VERSION) {
+                    return util::Error{strprintf(_("Can't spend unconfirmed version 3 pre-selected input with a version %d tx"), coin_control.m_version)};
+                } else if (coin_control.m_version == TRUC_VERSION && parent_tx.tx->version != TRUC_VERSION) {
+                    return util::Error{strprintf(_("Can't spend unconfirmed version %d pre-selected input with a version 3 tx"), parent_tx.tx->version)};
+                }
+            }
```
While the suggestions is shorter, I don't think this is a thorough as the original implementation. I'm not sure if transaction version 1 could cause an issue here - this suggestion can be ignored.

rkrux commented on test/functional/wallet_v3_txs.py:
```diff
@@ -0,0 +1,553 @@
+        inputs=[]
+        outputs = {self.bob.getnewaddress() : 1.999}
+        bob_txid = self.send_tx(self.bob, inputs, outputs, 3)
+        # alice spends both of her utxos, replacing bob's tx
+        outputs = {self.charlie.getnewaddress() : alice_v2_unspent['amount'] + alice_unspent['amount'] - Decimal(0.00005120)}
+        alice_txid = self.send_tx(self.alice, [alice_v2_unspent, alice_unspent], outputs, 3)
```
Ok, I debugged these transactions. 

In the `mempool_conflicts_removed_when_v3_conflict_removed` test, I can understand that the latter Alice tx has higher fees than the former Bob tx due to which Alice's tx sibling-evicts Bob's tx.

What's not clear to me is that why in the `v3_conflict_removed_from_mempool` test the latter Bob tx can't sibling-evict the former Alice's tx even though it has higher fees (and instead the "insufficient funds" is thrown)?

w0xlt reviewed (commented):
There's repeated logic for iterating through parent outputs and updating sibling transaction states in both `transactionAddedToMempool` and `transactionRemovedFromMempool`.

The below patch can improve this, using the same patter in both functions. If the logic for handling TRUC sibling conflicts needs to be changed in the future, it only needs to be updated in one place.

```diff
diff --git a/src/wallet/wallet.cpp b/src/wallet/wallet.cpp
index 28a646ef4b..c8e5e3a8e8 100644
--- a/src/wallet/wallet.cpp
+++ b/src/wallet/wallet.cpp
@@ -1374,6 +1374,23 @@ bool CWallet::TransactionCanBeAbandoned(const Txid& hashTx) const
     return wtx && !wtx->isAbandoned() && GetTxDepthInMainChain(*wtx) == 0 && !wtx->InMempool();
 }
 
+void CWallet::UpdateTrucSiblingConflicts(const CWalletTx& parent_wtx, const Txid& child_txid, bool add_conflict)
+{
+    AssertLockHeld(cs_wallet);
+    // Find all other txs in our wallet that spend utxos from this parent
+    // so that we can mark/unmark them as mempool-conflicted by this tx.
+    for (size_t i = 0; i < parent_wtx.tx->vout.size(); i++) {
+        for (auto range = mapTxSpends.equal_range(COutPoint(parent_wtx.tx->GetHash(), i)); range.first != range.second; range.first++) {
+            const Txid& sibling_txid = range.first->second;
+            // Skip the child tx itself
+            if (sibling_txid == child_txid) continue;
+            RecursiveUpdateTxState(/*batch=*/nullptr, sibling_txid, [&child_txid, add_conflict](CWalletTx& wtx) EXCLUSIVE_LOCKS_REQUIRED(cs_wallet) {
+                return add_conflict ? (wtx.mempool_conflicts.insert(child_txid).second ? TxUpdate::CHANGED : TxUpdate::UNCHANGED)
+                                    : (wtx.mempool_conflicts.erase(child_txid) ? TxUpdate::CHANGED : TxUpdate::UNCHANGED);
+            });
+        }
+    }
+}
+
 void CWallet::MarkInputsDirty(const CTransactionRef& tx)
 {
     for (const CTxIn& txin : tx->vin) {
@@ -1407,20 +1424,10 @@ void CWallet::transactionAddedToMempool(const CTransactionRef& tx) {
                 CWalletTx& parent_wtx = parent_it->second;
                 if (parent_wtx.isUnconfirmed()) {
                     parent_wtx.truc_child_in_mempool = tx->GetHash();
-                    // Find all other txs in our wallet that spend utxos from this parent
-                    // so that we can mark them as mempool-conflicted by this new tx.
                     // Even though these siblings do not spend the same utxos, they can't
                     // be present in the mempool at the same time because of TRUC policy rules
-                    for (long unsigned int i = 0; i < parent_wtx.tx->vout.size(); i++) {
-                        for (auto range = mapTxSpends.equal_range(COutPoint(parent_wtx.tx->GetHash(), i)); range.first != range.second; range.first++) {
-                            const Txid& sibling_txid = range.first->second;
-                            // Skip the recently added tx
-                            if (sibling_txid == txid) continue;
-                            RecursiveUpdateTxState(/*batch=*/nullptr, sibling_txid, [&txid](CWalletTx& parent_wtx) EXCLUSIVE_LOCKS_REQUIRED(cs_wallet) {
-                                return parent_wtx.mempool_conflicts.insert(txid).second ? TxUpdate::CHANGED : TxUpdate::UNCHANGED;
-                            });
-                        }
-                    }
+                    UpdateTrucSiblingConflicts(parent_wtx, txid, /*add_conflict=*/true);
                 }
             }
         }
@@ -1489,15 +1496,7 @@ void CWallet::transactionRemovedFromMempool(const CTransactionRef& tx, MemPoolRe
                 CWalletTx& parent_wtx = parent_it->second;
                 if (parent_wtx.truc_child_in_mempool == tx->GetHash()) {
                     parent_wtx.truc_child_in_mempool = std::nullopt;
-                    // Find all wallet transactions that spend utxos from the same parent_wtx
-                    for (long unsigned int i = 0; i < parent_wtx.tx->vout.size(); i++) {
-                        for (auto range = mapTxSpends.equal_range(COutPoint(parent_wtx.tx->GetHash(), i)); range.first != range.second; range.first++) {
-                            const Txid& spent_id = range.first->second;
-                            RecursiveUpdateTxState(/*batch=*/nullptr, spent_id, [&txid](CWalletTx& parent_wtx) EXCLUSIVE_LOCKS_REQUIRED(cs_wallet) {
-                                return parent_wtx.mempool_conflicts.erase(txid) ? TxUpdate::CHANGED : TxUpdate::UNCHANGED;
-                            });
-                        }
-                    }
+                    UpdateTrucSiblingConflicts(parent_wtx, txid, /*add_conflict=*/false);
                 }
             }
         }
diff --git a/src/wallet/wallet.h b/src/wallet/wallet.h
index e056b68182..1e3a1e0e48 100644
--- a/src/wallet/wallet.h
+++ b/src/wallet/wallet.h
@@ -598,6 +598,9 @@ private:
     //! Cache of descriptor ScriptPubKeys used for IsMine. Maps ScriptPubKey to set of spkms
     std::unordered_map<CScript, std::vector<ScriptPubKeyMan*>, SaltedSipHasher> m_cached_spks;
 
+    //! Update mempool conflicts for TRUC sibling transactions
+    void UpdateTrucSiblingConflicts(const CWalletTx& parent_wtx, const Txid& child_txid, bool add_conflict) EXCLUSIVE_LOCKS_REQUIRED(cs_wallet);
+
 public:
     /*
      * Main wallet lock.
```

ishaanam committed:
wallet: don't return utxos from multiple truc txs in AvailableCoins

ishaanam commented on src/policy/policy.cpp:
```diff
@@ -98,7 +98,7 @@ bool IsStandard(const CScript& scriptPubKey, TxoutType& whichType)
 
```
Done

ishaanam commented on test/functional/wallet_v3_txs.py:
```diff
@@ -0,0 +1,553 @@
+        # whitelist peers to speed up tx relay / mempool sync
+        self.noban_tx_relay = True
+
+    def send_tx(self, from_node, inputs, outputs, version):
```
Done

ishaanam commented on test/functional/wallet_v3_txs.py:
```diff
@@ -0,0 +1,553 @@
+        # unconfirmed v3 tx to alice & bob
+        self.log.info("Test v3 transaction with conflicting sibling")
+
+        outputs = {self.alice.getnewaddress() : 2.0, self.bob.getnewaddress() : 2.0}
```
Done

ishaanam commented on test/functional/wallet_v3_txs.py:
```diff
@@ -0,0 +1,553 @@
+            raw_tx = from_node.fundrawtransaction(raw_tx, {'include_unsafe' : True})["hex"]
+        raw_tx = from_node.signrawtransactionwithwallet(raw_tx)["hex"]
+        txid = from_node.sendrawtransaction(raw_tx)
+        self.sync_mempools()
```
Done

ishaanam commented on test/functional/wallet_v3_txs.py:
```diff
@@ -0,0 +1,553 @@
+        assert_equal(self.alice.gettransaction(alice_txid)['mempoolconflicts'], [bob_txid])
+
+        self.log.info("Test that re-submitting Alice's transaction with a higher fee removes bob's tx as a mempool conflict")
+        outputs = {self.alice.getnewaddress() : alice_unspent['amount'] - Decimal(0.00030120)}
```
Done

ishaanam commented on test/functional/wallet_v3_txs.py:
```diff
@@ -0,0 +1,553 @@
+        # alice fee-bumps her tx so it only spends the v2 utxo
+        outputs = {self.charlie.getnewaddress() : alice_v2_unspent['amount'] - Decimal(0.00015120)}
+        self.send_tx(self.alice, [alice_v2_unspent], outputs, 2)
+        # bob's tx now has non conflicts and can be rebroadcast
```
Done

ishaanam commented on test/functional/wallet_v3_txs.py:
```diff
@@ -0,0 +1,553 @@
+            assert_equal(0, self.alice.getbalances()["mine"]["trusted"])
+            assert_equal(0, self.bob.getbalances()["mine"]["trusted"])
+            assert_equal(0, self.alice.getbalances()["mine"]["immature"])
+            assert_equal(0, self.bob.getbalances()["mine"]["immature"])
```
Done

ishaanam commented on test/functional/wallet_v3_txs.py:
```diff
@@ -0,0 +1,553 @@
+)
+
+def cleanup(func):
+    def wrapper(self, *args):
```
Done

ishaanam commented on test/functional/wallet_v3_txs.py:
```diff
@@ -0,0 +1,553 @@
+    @cleanup
+    def v3_tx_evicted_from_mempool_by_sibling(self):
+        # unconfirmed v3 tx to alice & bob
+        self.log.info("Test v3 transaction with conflicting sibling")
```
Done

ishaanam commented on test/functional/wallet_v3_txs.py:
```diff
@@ -0,0 +1,553 @@
+
+    @cleanup
+    def truc_tx_with_conflicting_sibling_change(self):
+        # unconfirmed v3 tx to alice & bob
```
Done

glozow commented on src/wallet/spend.cpp:
```diff
@@ -316,6 +325,9 @@ CoinsResult AvailableCoins(const CWallet& wallet,
     CoinsResult result;
+    // track unconfirmed truc outputs separately if we are tracking trucness
+    CoinsResult unconfirmed_truc_coins;
+    std::unordered_map<uint256, CAmount, SaltedTxidHasher> truc_txid_by_value;
```
should use `Txid` when possible

ishaanam replied:
Updated here and later in the function

glozow commented on src/wallet/spend.cpp:
```diff
@@ -468,6 +501,21 @@ CoinsResult AvailableCoins(const CWallet& wallet,
     }
 
+    if (params.check_version_trucness && unconfirmed_truc_coins.Size() > 0) {
+        auto highest_value_truc_tx = std::max_element(truc_txid_by_value.begin(), truc_txid_by_value.end(), [](const auto& tx1, const auto& tx2){
+                return tx1.second < tx2.second;
+                });
+
+        const uint256& truc_txid = highest_value_truc_tx->first;
+        for (const auto& [type, outputs] : unconfirmed_truc_coins.coins) {
+            for (const COutput& output : outputs) {
+                if (output.outpoint.hash == truc_txid) {
+                        result.Add(type, output);
+                }
+            }
+        }
```
Do you need to track all of them or can you just keep a `pair<Txid, CAmount>`?

ishaanam commented on src/wallet/rpc/spend.cpp:
```diff
@@ -704,6 +705,12 @@ CreatedTransactionResult FundTransaction(CWallet& wallet, const CMutableTransact
     }
 
+    if (tx.version == TRUC_VERSION) {
+        if (!coinControl.m_max_tx_weight.has_value() || coinControl.m_max_tx_weight.value() > TRUC_MAX_VSIZE * WITNESS_SCALE_FACTOR) {
```
Done

ishaanam commented on src/wallet/spend.cpp:
```diff
@@ -925,11 +925,17 @@ util::Result<SelectionResult> AutomaticCoinSelection(const CWallet& wallet, Coin
             if (it == filtered_groups.end()) continue;
+            if (temp_selection_params.m_version == TRUC_VERSION && (select_filter.filter.conf_mine == 0 || select_filter.filter.conf_theirs == 0)) {
+                if (temp_selection_params.m_max_tx_weight > (TRUC_CHILD_MAX_VSIZE * WITNESS_SCALE_FACTOR)) {
+                    temp_selection_params.m_max_tx_weight = TRUC_CHILD_MAX_VSIZE * WITNESS_SCALE_FACTOR;
```
Done

ishaanam commented on src/wallet/spend.cpp:
```diff
@@ -925,11 +925,17 @@ util::Result<SelectionResult> AutomaticCoinSelection(const CWallet& wallet, Coin
             if (it == filtered_groups.end()) continue;
+            if (temp_selection_params.m_version == TRUC_VERSION && (select_filter.filter.conf_mine == 0 || select_filter.filter.conf_theirs == 0)) {
+                if (temp_selection_params.m_max_tx_weight > (TRUC_CHILD_MAX_VSIZE * WITNESS_SCALE_FACTOR)) {
+                    temp_selection_params.m_max_tx_weight = TRUC_CHILD_MAX_VSIZE * WITNESS_SCALE_FACTOR;
```
I've updated the name to `updated_selection_params`. I agree that updating `coin_selection_params` directly would be outside the scope of this pr.

ishaanam commented on src/wallet/wallet.cpp:
```diff
@@ -1238,6 +1238,22 @@ bool CWallet::TransactionCanBeAbandoned(const Txid& hashTx) const
     const CWalletTx* wtx = GetWalletTx(hashTx);
     return wtx && !wtx->isAbandoned() && GetTxDepthInMainChain(*wtx) == 0 && !wtx->InMempool();
 }
+void CWallet::UpdateTrucSiblingConflicts(const CWalletTx& parent_wtx, const Txid& child_txid, bool add_conflict) EXCLUSIVE_LOCKS_REQUIRED(cs_wallet)
```
Done

glozow commented on src/wallet/spend.cpp:
```diff
@@ -468,6 +498,19 @@ CoinsResult AvailableCoins(const CWallet& wallet,
     }
 
+    if (params.check_version_trucness && unconfirmed_truc_coins.size() > 0) {
+        auto highest_value_truc_tx = std::max_element(truc_txid_by_value.begin(), truc_txid_by_value.end(), [](const auto& tx1, const auto& tx2){
```
Do I understand correctly that the approach here is to just return 1 unconfirmed TRUC coin from `AvailableCoins`, i.e. the one with the highest value? That seems like it could weaken coin selection quality a little bit - is there a way to encode these kinds of restrictions (once coin X is selected, ABC are off the table) into the coin selection process instead?

This is probably rare in practice (I think people would only use unconfirmed TRUC when CPFPing or really desperate), but it would be helpful to provide justification here.

glozow reviewed (commented):
reACK 5c087f58245cfba6b61691ece5387eccf8a1a5f9

w0xlt reviewed (commented):
ACK https://github.com/bitcoin/bitcoin/pull/32896/commits/5c087f58245cfba6b61691ece5387eccf8a1a5f9

rkrux reviewed (approved):
lgtm ACK 5c087f58245cfba6b61691ece5387eccf8a1a5f9

Thanks for taking this over the finish line. To implement the TRUC conditions in the wallet, I've observed that the related changes are done in the following control sites of selecting coins while building the transaction via different corresponding RPCs:
1. `FetchSelectedInputs`: to ensure the parent-child TRUC versions are in agreement.
2. `AvailableCoins`: to ensure the parent-child TRUC versions are in agreement, to ensure the TRUC ancestor/descendant limits are adhered to.
3. `AutomaticCoinSelection`: to ensure the TRUC transaction weight criterion are met.
4. Notifications - `transactionAddedToMempool`, `transactionRemovedFromMempool`: to ensure transaction mempool conflicts are updated and unconfirmed TRUC child is tracked.

The several functional tests also give more confidence.

The couple nits below can be done if there is any other need to retouch the PR.

ishaanam commented on src/wallet/spend.cpp:
```diff
@@ -470,8 +473,15 @@ CoinsResult AvailableCoins(const CWallet& wallet,
+        } else {
+            result.Add(GetOutputType(type, is_from_p2sh),
+                       COutput(outpoint, output, nDepth, input_bytes, spendable, solvable, tx_safe, wtx.GetTxTime(), tx_from_me, feerate));
+        }
```
Done

ishaanam commented on src/wallet/spend.cpp:
```diff
@@ -488,6 +498,24 @@ CoinsResult AvailableCoins(const CWallet& wallet,
     }
 
+    // Return the unconfirmed TRUC coin, or multiple TRUC coins from
+    // the same transaction, that have the highest value. This could
```
Done

ishaanam commented on src/wallet/rpc/spend.cpp:
```diff
@@ -1369,6 +1371,7 @@ RPCHelpMan sendall()
                         {"send_max", RPCArg::Type::BOOL, RPCArg::Default{false}, "When true, only use UTXOs that can pay for their own fees to maximize the output amount. When 'false' (default), no UTXO is left behind. send_max is incompatible with providing specific inputs."},
                         {"minconf", RPCArg::Type::NUM, RPCArg::Default{0}, "Require inputs with at least this many confirmations."},
                         {"maxconf", RPCArg::Type::NUM, RPCArg::Optional::OMITTED, "Require inputs with at most this many confirmations."},
+                        {"version", RPCArg::Type::NUM, RPCArg::Default{DEFAULT_WALLET_TX_VERSION}, "Transaction version"},
```
Done

achow101 commented on src/wallet/spend.cpp:
```diff
@@ -488,6 +498,23 @@ CoinsResult AvailableCoins(const CWallet& wallet,
+        const Txid& truc_txid = highest_value_truc_tx->first;
+        for (const auto& [type, output] : unconfirmed_truc_coins) {
+            if (output.outpoint.hash == truc_txid) {
+                    result.Add(type, output);
```
In c5a2d080116270ecd0414c14eb412fa30eaaedaf "wallet: don't return utxos from multiple truc txs in AvailableCoins"

nit: extra indent

achow101 reviewed (commented):
ACK 5c8bf7b39e9bffba7c4d5778b56b1ebe72f5ea1b

I'm not fully sold on the approach of choosing the highest value TRUC UTXO, but don't have any ideas on how to change that for this PR. Maybe @murchandamus has some thoughts on that.

rkrux reviewed (approved):
ACK 5c8bf7b39e9bffba7c4d5778b56b1ebe72f5ea1b

```
git range-diff 5c087f5...5c8bf7b
```

murchandamus commented:
> I'm not fully sold on the approach of choosing the highest value TRUC UTXO, but don't have any ideas on how to change that for this PR. Maybe @murchandamus has some thoughts on that.

Been pondering this a bit today. 

- In the first pass, we will always try to fund transactions only with confirmed UTXOs. If we are just requesting a new TRUC transaction, it will always get funded by confirmed UTXOs if we have sufficient confirmed funds.
- If we are deliberately creating a child transaction, it’s probably because we want to CPFP an existing TRUC transaction and that already restricts the unconfirmed UTXOs to be created by the specific parent transaction. In that case, only other outputs from the same transactions are other eligible unconfirmed inputs.
- This leaves only the case in which we are creating a TRUC transaction, but we do not have sufficient confirmed funds to pay for it. In that case, we can use one or several unconfirmed UTXOs from a single transaction in addition to any number of confirmed UTXOs.

We do have the `nDepth` member on `COutput`, so if we have several unconfirmed single outputs from transactions, we could keep track of whether we have selected an unconfirmed UTXO. Whenever we were to add a second unconfirmed UTXO, we could keep it or replace it with new one, depending on which has the greater effective value instead of adding it. Perhaps, just in case of creating TRUC transactions, we would create a new type of `OutputGroups` based on common parent transaction instead of common output script in that context. Because we can only use one `OutputGroup` that contains unconfirmed UTXOs, we could generate all subsets of a transaction’s outputs as separate OutputGroups, and thereby have those unconfirmed UTXOs appear multiple times in different configurations. We could only use that approach if there is a small number (<10?) of outputs to ourselves on the transaction as power sets grow exponentially. Either way, it  might be easier to have a dedicated algorithm for TRUC child transactions, that tracks the two pots separately, but all of that seems very complicated.

Given that it should be fairly uncommon to have all of your funds in flight and the complexity of alternatives, using the largest UTXO or group of UTXOs feels viable to me, although it could be improved in the future.

hebasto commented on src/wallet/spend.cpp:
```diff
@@ -282,6 +283,14 @@ util::Result<PreSelectedInputs> FetchSelectedInputs(const CWallet& wallet, const
             }
+            const CWalletTx& parent_tx = txo->GetWalletTx();
+            if (wallet.GetTxDepthInMainChain(parent_tx) == 0) {
+                if (parent_tx.tx->version == TRUC_VERSION && coin_control.m_version != TRUC_VERSION) {
+                    return util::Error{strprintf(_("Can't spend unconfirmed version 3 pre-selected input with a version %d tx"), coin_control.m_version)};
+                } else if (coin_control.m_version == TRUC_VERSION && parent_tx.tx->version != TRUC_VERSION) {
+                    return util::Error{strprintf(_("Can't spend unconfirmed version %d pre-selected input with a version 3 tx"), parent_tx.tx->version)};
+                }
```
Translations followup in https://github.com/bitcoin/bitcoin/pull/33193.
